core/integration/module_cli_test.go

```diff
@@ -1300,3 +1322,266 @@ func (f *Foo) ContainerEcho(ctx context.Context, input string) (string, error) {
 		}
 	})
 }
+
+func (CLISuite) TestDaggerUpdate(ctx context.Context, t *testctx.T) {
+	var daggerjsonContains = func(list []string) func(t *testctx.T, daggerjson string) {
+		return func(t *testctx.T, daggerjson string) {
+			for _, s := range list {
+				require.Contains(t, daggerjson, s)
+			}
+		}
+	}
+
+	var daggerjsonNotContains = func(list []string) func(t *testctx.T, daggerjson string) {
+		return func(t *testctx.T, daggerjson string) {
+			for _, s := range list {
+				require.NotContains(t, daggerjson, s)
+			}
+		}
+	}
+
+	testcases := []struct {
+		name          string
+		initCmd       []string
+		installCmds   [][]string
+		verifySetup   func(t *testctx.T, daggerjson string)
+		updateCmd     []string
+		verifyFns     []func(t *testctx.T, daggerjson string)
+		expectedError string
+	}{
+		{
+			name:    "existing dep has version, update cmd has version",
+			initCmd: []string{"init", "--sdk=go", "--name=foo", "--source=."},
+			installCmds: [][]string{
+				{"install", "github.com/shykes/daggerverse/docker@v0.4.1"},
+			},
+			verifySetup: daggerjsonContains([]string{`"github.com/shykes/daggerverse/docker@docker/v0.4.1"`}),
+			updateCmd:   []string{"update", "github.com/shykes/daggerverse/docker@v0.4.2"},
+			verifyFns: []func(t *testctx.T, daggerjson string){
+				daggerjsonContains([]string{`"github.com/shykes/daggerverse/docker@docker/v0.4.2"`}),
+				daggerjsonNotContains([]string{`github.com/shykes/daggerverse/docker@docker/v0.4.1`}),
+			},
+			expectedError: "",
```

Comment:
- `initCmd` is always the same - so can be removed.
- We don't need `verifySetup` - testing this behavior should actually be done in `TestDaggerInstall` (we shouldn't duplicate the testing)
- `daggerjsonContains` is kinda neat, but I think we could just have `contains` and `notContains` properties right?
- The default value of `expectedErr` is `""` anyways - so we don't need to manually set it (same with all other zero values)

---

Just a heads-up - if the tests start to feel tricky to get all into one giant table test, that's not an issue - you can split up testing into subtests (e.g. you could one that tests the `pin` updating, another that tests upgrading to a specified branch/tag, etc). Not required, but that's another option if the table tests start to feel too large.